packages/context/src/binding.ts

```diff
@@ -148,7 +157,7 @@ type ValueGetter<T> = (
  * Binding represents an entry in the `Context`. Each binding has a key and a
  * corresponding value getter.
  */
-export class Binding<T = BoundValue> {
+export class Binding<T = BoundValue> extends EventEmitter {
```

Comment:
I am concerned about the performance implications of turning every binding into an event emitter. I am expecting that consumers of these events will want to listen on Context instances, not on individual bindings.

Have you considered changing this implementation to emit the events on the Context object the binding is added to? If the binding is not added to any Context object yet, then I am arguing that the events can be silently discarded.

I am also surprised that we did not these Binding-level events when implementing [Context Observer](https://loopback.io/doc/en/lb4/Context.html#context-observers). Does it mean that Context observers were not able to be notified about these kinds of changes before? Did we perhaps use a different mean to achieve this functionality for context observers?

Reply:
At this moment, the same binding instance can be added to multiple context objects. If we decide to treat each binding to be optionally owned by one context, we can emit such events on the owning context.

When we implemented ContextObserver - which is async, we choose to process `bind` and `unbind` events at  a late cycle to work around the issue that bindings can be changed. Now with binding events, we can probably revisit the implementation to track `binding.changed` events.

Reply:
IMO, setting `Context` ref on `Binding` is similar as `Context` registering itself as a listener on `Binding` from performance overhead perspective.

Reply:
> At this moment, the same binding instance can be added to multiple context objects.

Somehow I thought that a Binding is already holding a reference to the single owning Context. I check the source code and see that it does not. As you wrote, it's possible to add the same binding to multiple contexts, therefore I agree with your originally proposed design, where each Binding is an EventEmitter 👍 

> When we implemented ContextObserver - which is async, we choose to process `bind` and `unbind` events at a late cycle to work around the issue that bindings can be changed. Now with binding events, we can probably revisit the implementation to track `binding.changed` events.

Make sense. I'd like to revisit this part sooner rather than later, to ensure consistency.